src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+;;; can't be loaded until its Table is defined, and the Table can't be loaded without the Database. These dependencies
+;;; are derived from the serialized form, and then loaded recursively in postorder.
+(defmulti serdes-dependencies
+  "Given an entity map AS DECODED FROM YAML, returns a (possibly empty) list of its dependencies, where each dependency
+  is represented by `(or entity_id identity-hash)` of the target entity.
+
+  This is mostly part of the deserialization process, so it's based on the serialized map and not on an IModel instance.
+
+  Default implementation returns an empty vector, so only models that have dependencies need to implement this."
+  :serdes_type)
+
+(defmethod serdes-dependencies :default [_]
+  [])
+
+(p.types/defprotocol+ ISerializable
+  ;;; Serialization side
+  (serialize-all
+    [this user-or-nil]
+    "Serializes all relevant instances of this model.
+    The return value is a reducible sequence of [file-path entity-map] pairs, possibly empty.
+    There can be a single file for all the entities (eg. settings) or one file per entity.
+
+    The default implementation assumes there will be many files:
+    - It calls (serialize-query this user-or-nil) to get the (lazy, reducible) set of entities to serialize.
+    - For each entity, it calls `(serialize-one e)` and expects a `[file-path {...}]` pair.
+    - These pairs are returned as a reducible stream.
+
+    To serialize as a single file, override this to do something else. (Then `serialize-query` and `serialize-one` are
+    never called and can be stubs.)")
+
+  (serialize-query
+    [this user-or-nil]
+    "Performs the select query, possibly filtered, for all the entities of this type that should be serialized.
+    Returns the result of the db/select-reducible.
+    Defaults to a naive `db/select-reducible` for the entire model, ignoring the optional user argument.
+
+    Only called by `serialize-all`, either the default implementation or (optionally) by a custom one.
+    This exists to be an easy override point, when the default `serialize-all` is good but you need to filter
+    the returned set, eg. by dropping archived entities.")
+
+  (serialize-one
+    [this]
+    "Serializes a single entity into a YAML-friendly map, with its filename.
+    Returns a [file-name {entity map...}] pair.
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
```

Comment:
It would be SUPER DUPER helpful to document every variable in methods like this that we expect to implement in a few dozen places. I do not know what `local` means here. I'm guessing  it's a boolean value (in which case I would personally name it `local?` but that's a style preference hill that I'm not willing to die on)? 

The only way to figure out what `local` means/is used for is to go read implementations of the method, which is definitely a sign the docstring isn't detailed enough. Ideally we should know what a function does, what it returns, and what all the parameters should be/are used for just by reading the docstring.

Reply:
Docs are a lot better in the rebuilt PR.